itest/utils.go

```diff
@@ -735,6 +739,182 @@ func ManualMintSimpleAsset(t *harnessTest, lndNode *node.HarnessNode,
+// calls the external signer callback to sign the group virtual PSBT.
+func MintAssetExternalSigner(t *harnessTest, tapNode *tapdHarness,
+	assetReqs []*mintrpc.MintAssetRequest,
+	externalSignerCallback ExternalSigCallback) []*taprpc.Asset {
+
+	BuildMintingBatch(t.t, tapNode, assetReqs)
+
+	// Fund mint batch with BTC.
+	ctxb := context.Background()
+	ctxt, cancel := context.WithTimeout(ctxb, defaultWaitTimeout)
+
+	fundResp, err := tapNode.FundBatch(ctxt, &mintrpc.FundBatchRequest{})
+	require.NoError(t.t, err)
+
+	// Cancel the context for the fund request call.
+	cancel()
+
+	require.NotEmpty(t.t, fundResp.Batch)
+	require.Equal(
+		t.t, mintrpc.BatchState_BATCH_STATE_PENDING,
+		fundResp.Batch.Batch.State,
+	)
+	require.Len(t.t, fundResp.Batch.UnsealedAssets, 1)
+
+	// Pass unsealed assets to external signer callback to sign the group
+	// virtual PSBT.
+	callbackRes := externalSignerCallback(fundResp.Batch.UnsealedAssets)
+
+	// Extract group witness from signed PSBTs.
+	var groupWitnesses []*taprpc.GroupWitness
+	for idx := range callbackRes {
+		res := callbackRes[idx]
+		signedPsbt := res.SignedPsbt
+		genesisAssetID := res.AssetID
+
+		// Sanity check signed PSBT.
+		require.Len(t.t, signedPsbt.Inputs, 1)
+		require.Len(t.t, signedPsbt.Outputs, 1)
+
+		// Extract witness from signed PSBT.
+		witnessStack, err := DeserializeWitnessStack(
+			signedPsbt.Inputs[0].FinalScriptWitness,
+		)
+		if err != nil {
+			log.Fatalf("Failed to deserialize witness stack: %v",
+				err)
+		}
+
+		groupWitness := taprpc.GroupWitness{
+			GenesisId: genesisAssetID[:],
+			Witness:   witnessStack,
+		}
+
+		groupWitnesses = append(groupWitnesses, &groupWitness)
+	}
+
+	// Seal the batch with the group witnesses.
+	ctxt, cancel = context.WithTimeout(ctxb, defaultWaitTimeout)
+
+	sealReq := mintrpc.SealBatchRequest{
+		GroupWitnesses: groupWitnesses,
+	}
+	sealResp, err := tapNode.SealBatch(ctxt, &sealReq)
+	require.NoError(t.t, err)
+
+	// Cancel the context for the seal request call.
+	cancel()
+
+	require.NotEmpty(t.t, sealResp.Batch)
+
+	// With the batch sealed successfully, we can now finalize it and
+	// broadcast the anchor TX.
+	ctxt, cancel = context.WithCancel(context.Background())
+	defer cancel()
+	stream, err := tapNode.SubscribeMintEvents(
+		ctxt, &mintrpc.SubscribeMintEventsRequest{},
+	)
+	require.NoError(t.t, err)
+	sub := &EventSubscription[*mintrpc.MintEvent]{
+		ClientEventStream: stream,
+		Cancel:            cancel,
+	}
+
+	batchTXID, batchKey := FinalizeBatchUnconfirmed(
+		t.t, t.lndHarness.Miner().Client, tapNode, assetReqs,
+	)
+	batchAssets := ConfirmBatch(
+		t.t, t.lndHarness.Miner().Client, tapNode, assetReqs, sub,
+		batchTXID, batchKey,
+	)
+
+	return batchAssets
+}
+
+// DeserializeWitnessStack deserializes a serialized witness stack into a
+// [][]byte.
+//
+// TODO(ffranr): Do we already have this function somewhere else? Move to btcd?
+func DeserializeWitnessStack(serialized []byte) ([][]byte, error) {
```

Comment:
Could probably use `asset.TxWitnessDecoder()` somehow.

Reply:
I will reformulate in https://github.com/lightninglabs/taproot-assets/pull/1295
where we will extract the witness stack from the signed PSBT in tapd.